code/modules/mob/living/carbon/xenomorph/castes/behemoth/abilities_behemoth.dm

```diff
@@ -2,1474 +2,1217 @@
+	var/creation_limit = 1
+
+/datum/action/ability/activable/xeno/earth_riser/give_action(mob/living/L)
 	. = ..()
-	switch(direction)
-		if(NORTH)
-			pixel_x += which_step? -12 : 12
-			pixel_y += 17
-		if(SOUTH)
-			pixel_x += which_step? 12 : -12
-			pixel_y -= 24
-		if(WEST)
-			pixel_x -= 16
-			pixel_y += which_step? -10 : 10
-		if(EAST)
-			pixel_x += 20
-			pixel_y += which_step? 10 : -10
+	LAZYINITLIST(active_pillars)
+	RegisterSignal(xeno_owner, COMSIG_MOB_DEATH, PROC_REF(clear_pillars))
+	visual_references[VREF_MUTABLE_EARTH_PILLAR] = mutable_appearance(null, null, ACTION_LAYER_MAPTEXT, appearance_flags = RESET_COLOR)
 
-/obj/effect/temp_visual/behemoth/landslide/dust
-	icon = 'icons/effects/effects.dmi'
-	icon_state = "landslide_dust"
-	layer = MOB_BELOW_PIGGYBACK_LAYER
-	duration = 1.1 SECONDS
+/datum/action/ability/activable/xeno/earth_riser/remove_action(mob/living/L)
+	clear_pillars()
+	return ..()
+
+/datum/action/ability/activable/xeno/earth_riser/clean_action()
+	clear_pillars()
+	return ..()
 
-/obj/effect/temp_visual/behemoth/landslide/dust/Initialize(mapload, direction, which_step)
+/datum/action/ability/activable/xeno/earth_riser/update_button_icon()
 	. = ..()
-	adjust_offsets(direction, which_step)
-	do_animation(direction)
+	if(!. || !visual_references[VREF_MUTABLE_EARTH_PILLAR])
+		return
+	var/mutable_appearance/counter = visual_references[VREF_MUTABLE_EARTH_PILLAR]
+	button.cut_overlay(counter)
+	if(!creation_limit)
+		return
+	var/condition = creation_limit > 2 ? TRUE : FALSE
+	counter.icon = condition ? null : action_icon
+	counter.icon_state = condition ? null : "pillar_counter_[creation_limit][LAZYLEN(active_pillars)]" // f.ex pillar_counter_11 means we have 1 pillar as a limit and 1 active pillar
+	counter.pixel_x = condition ? 16 : initial(counter.pixel_x)
+	counter.pixel_y = condition ? -4 : initial(counter.pixel_y)
+	counter.maptext = condition ? MAPTEXT("["[LAZYLEN(active_pillars)]/[creation_limit]"]") : null
+	visual_references[VREF_MUTABLE_EARTH_PILLAR] = counter
+	button.add_overlay(counter)
+
+// Handles ability checks instead, mostly because we need the final target in the event of target correction.
+// We also check for plasma costs and cooldowns here because some parts of this ability ignore them.
+/datum/action/ability/activable/xeno/earth_riser/use_ability(atom/target)
+	. = ..()
+	if(!xeno_owner.incapacitated())
+		if(xeno_owner.held_pillar) // If we're holding an Earth Pillar, then we want to throw it.
+			if(xeno_owner.plasma_stored < EARTH_RISER_THROW_COST)
+				xeno_owner.balloon_alert(xeno_owner, "Need [EARTH_RISER_THROW_COST - xeno_owner.plasma_stored] more plasma")
+				return
+			if(!action_cooldown_finished())
+				xeno_owner.balloon_alert(xeno_owner, "Wait [cooldown_remaining()] seconds!")
+				return
+			if(!line_of_sight(xeno_owner, target, WORLD_VIEW_NUM))
+				xeno_owner.balloon_alert(xeno_owner, "No line of sight with [target]")
```

Comment:
generally don't use names in balloon alerts because they can become huge and annoying. They should be as short as possible while conveying the useful information.